pyro/distributions/transforms/block_naf.py

```diff
@@ -0,0 +1,234 @@
+    """
+    return torch.logsumexp(A.unsqueeze(-1) + B.unsqueeze(-3), dim=-2)
+
+
+@copy_docs_from(TransformModule)
+class BlockNAFFlow(TransformModule):
+    """
+    An implementation of Block Neural Autoregressive Flow (block-NAF) (De Cao et al., 2019). Block-NAF uses a
+    similar transformation to deep dense NAF, building the autoregressive NN into the structure of the flow,
+    in a sense.
+
+    Together with `TransformedDistribution` this provides a way to create richer variational approximations.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> naf = BlockNAFFlow(input_dim=10)
+    >>> pyro.module("my_naf", naf)  # doctest: +SKIP
+    >>> naf_dist = dist.TransformedDistribution(base_dist, [naf])
+    >>> naf_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse operation is not implemented. This would require numerical inversion, e.g., using a
+    root finding method - a possibility for a future implementation.
+
+    :param input_dim: The dimensionality of the input and output variables.
+    :type input_dim: int
+    :param hidden_factors: Hidden layer i has hidden_factors[i] hidden units per input dimension. This
+        corresponds to both :math:`a` and :math:`b` in De Cao et al. (2019). The elements of hidden_factors
+        must be integers.
+    :type hidden_factors: list
+    :param activation: Activation function to use. One of 'ELU', 'LeakyReLU', 'sigmoid', or 'tanh'.
+    :type activation: string
+    :param residual: Type of residual connections to use. Choices are "None", "normal" for
+        :math:`\\mathbf{y}+f(\\mathbf{y})`, and "gated" for :math:`\\alpha\\mathbf{y} + (1 - \\alpha\\mathbf{y})`
+        for learnable parameter :math:`\\alpha`.
+    :type residual: string
+
+    References:
+
+    Block Neural Autoregressive Flow [arXiv:1904.04676]
+    Nicola De Cao, Ivan Titov, Wilker Aziz
+    """
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+    autoregressive = True
+
+    def __init__(self, input_dim, hidden_factors=[8, 8], activation='tanh', residual=None):
+        super(BlockNAFFlow, self).__init__(cache_size=1)
+
+        if any([h < 1 for h in hidden_factors]):
+            raise ValueError('Hidden factors, {}, must all be >= 1'.format(hidden_factors))
+
+        if residual not in [None, 'normal', 'gated']:
+            raise ValueError('Invalid value {} for keyword argument "residual"'.format(residual))
+
+        # Mix in activation function methods
+        name_to_mixin = {'ELU': ELUMixin, 'LeakyReLU': LeakyReLUMixin, 'sigmoid': SigmoidalMixin, 'tanh': TanhMixin}
+        if activation not in name_to_mixin:
+            raise ValueError('Invalid activation function "{}"'.format(activation))
+        self.f = types.MethodType(name_to_mixin[activation].f, self)
```

Comment:
@neerajprad can you please take a look at use of `type` and mixins here?

Reply:
I've removed the use of `types.MethodType` in favour of static methods as I couldn't get it to work on Python 2.7